pkg/orchestrator/housekeeping.go

```diff
@@ -0,0 +1,219 @@
+	"github.com/bacalhau-project/bacalhau/pkg/lib/validate"
+	"github.com/bacalhau-project/bacalhau/pkg/models"
+	"github.com/rs/zerolog/log"
+)
+
+const (
+	// DefaultHousekeepingWorkers is the default number of parallel workers for housekeeping tasks
+	DefaultHousekeepingWorkers = 3
+)
+
+type HousekeepingParams struct {
+	EvaluationBroker EvaluationBroker
+	JobStore         jobstore.Store
+	// Interval is the interval at which housekeeping tasks are run
+	Interval time.Duration
+	// Workers is the maximum number of parallel workers for housekeeping tasks
+	Workers int
+	// TimeoutBuffer is the buffer time to add to the execution timeout
+	// It is better that compute nodes timeout and report the failure before the orchestrator does.
+	// This buffer is added to the execution timeout to allow for this.
+	TimeoutBuffer time.Duration
+}
+
+type Housekeeping struct {
+	evaluationBroker EvaluationBroker
+	jobStore         jobstore.Store
+	interval         time.Duration
+	timeoutBuffer    time.Duration
+
+	workersSem chan struct{}
+	waitGroup  sync.WaitGroup
+	ctx        context.Context
+	cancel     context.CancelFunc
+	startOnce  sync.Once
+}
+
+func NewHousekeeping(params HousekeepingParams) (*Housekeeping, error) {
+	if params.Workers == 0 {
+		params.Workers = DefaultHousekeepingWorkers
+	}
+
+	// validate params
+	err := errors.Join(
+		validate.IsNotNil(params.EvaluationBroker, "evaluation broker cannot be nil"),
+		validate.IsNotNil(params.JobStore, "job store cannot be nil"),
+		validate.IsGreaterThanZero(params.Interval, "interval must be greater than zero"),
+		validate.IsGreaterThanZero(params.Workers, "workers must be greater than zero"),
+		validate.IsGreaterThanZero(params.TimeoutBuffer, "timeout buffer must be greater than zero"),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error validating housekeeping params. %w", err)
+	}
+
+	h := &Housekeeping{
+		evaluationBroker: params.EvaluationBroker,
+		jobStore:         params.JobStore,
+		interval:         params.Interval,
+		timeoutBuffer:    params.TimeoutBuffer,
+		workersSem:       make(chan struct{}, params.Workers),
+	}
+
+	return h, nil
+}
+
+// ShouldRun returns true if the housekeeping task should run.
+// This is just a placeholder for now until we introduce leader election or lease management for housekeeping
+// when we introduce more than one orchestrator.
+func (h *Housekeeping) ShouldRun() bool {
+	return true
+}
+
+// Start starts the housekeeping task
+func (h *Housekeeping) Start(ctx context.Context) {
+	h.startOnce.Do(func() {
+		h.ctx, h.cancel = context.WithCancel(ctx)
+		go h.runHousekeepingTasks()
+	})
+}
+
+func (h *Housekeeping) Stop(ctx context.Context) {
+	if h.cancel != nil {
+		h.cancel()
+	}
+
+	// wait for inflight housekeeping tasks to complete, or until the context is done
+	waitGroupDone := make(chan struct{})
+	go func() {
+		h.waitGroup.Wait()
+		close(waitGroupDone)
+	}()
+
+	select {
+	case <-waitGroupDone:
+	case <-ctx.Done():
+		h.waitGroup.Done()
```

Comment:
Blocking:
This isn't guaranteed to cause the above go routine to exit since several active go routines in `timeoutExecutions` may have added to it; `Done` only decrements the counter by 1. 
Further, this code may panic if the current count of the wait group is 0 and `Done` is called since the wait group panics if the count goes below zero: https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/sync/waitgroup.go;l=32

Reply:
Good catch. That shouldn't be there. Removed in https://github.com/bacalhau-project/bacalhau/pull/3974/commits/4527fcdac6e33cd4e08cee9f926e03318dc36ade